eng/common/scripts/job-matrix/README.md

```diff
@@ -0,0 +1,455 @@
+# Azure Pipelines Matrix Generator
+
+* [Usage in a pipeline](#usage-in-a-pipeline)
+* [Matrix config file syntax](#matrix-config-file-syntax)
+ * [Fields](#fields)
+    * [matrix](#matrix)
+    * [include](#include)
+    * [exclude](#exclude)
+    * [displayNames](#displaynames)
+    * [$IMPORT](#import)
+* [Matrix Generation behavior](#matrix-generation-behavior)
+    * [all](#all)
+    * [sparse](#sparse)
+    * [include/exclude](#includeexclude)
+    * [displayNames](#displaynames-1)
+    * [Filters](#filters)
+    * [NonSparseParameters](#nonsparseparameters)
+    * [Under the hood](#under-the-hood)
+* [Testing](#testing)
+
+
+This directory contains scripts supporting dynamic, cross-product matrix generation for azure pipeline jobs.
+It aims to replicate the [cross-product matrix functionality in github actions](https://docs.github.com/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-running-with-more-than-one-version-of-nodejs),
+but also adds some additional features like sparse matrix generation, cross-product includes and excludes, and programmable matrix filters.
+
+This functionality is made possible by the ability for the azure pipelines yaml to take a [dynamic variable as an input
+for a job matrix definition](https://docs.microsoft.com/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#multi-job-configuration) (see the code sample at the bottom of the linked section).
+
+## Usage in a pipeline
+
+In order to use these scripts in a pipeline, you must provide a config file and call the matrix creation script within a powershell job.
+
+For a single matrix, you can include the `/eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml` template in a pipeline (see /eng/common/scripts/job-matrix/samples/matrix-test.yml for a full working example):
+
+```
+jobs:
+  - template: /eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
+    parameters:
```

Comment:
Lets go ahead and update this to include the path to the job template as well.

Reply:
The path is correct here. The hierarchy mirrors the repos, so:

azure-sdk-tools: 
  /eng/common/scripts/job-matrix/samples/matrix-test.yml ->
    /eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml ->
      /eng/common/scripts/job-matrix/samples/matrix-job-sample.yml

azure-sdk-for-something: 
  /eng/common/pipelines/templates/stages/archetype-sdk-tests.yml ->
    /eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml ->
      /eng/common/pipelines/templates/jobs/archetype-sdk-tests-jobs.yml

In the README, I just copied a snippet from the matrix-test.yml file which is the top level pipeline file, and added a note to look at the full file.

Reply:
I was only requesting that the sample here illustrate passing the new JobTemplatePath parameter as well.

Reply:
Ahh, sorry, misunderstood!